### PR TITLE
[mlir][OpenMP] rewrite conversion of privatisation for omp.parallel

### LINKED ISSUE #106297: [flang][OpenMP] Parallel region failure
Here's a test program:
```
program bug
  use omp_lib
  implicit none
  integer ivar
  real :: table(10)
  !$OMP PARALLEL NUM_THREADS(3) PRIVATE(table)
    table = 1
    !$OMP PARALLEL NUM_THREADS(3) SHARED(table)
      !$OMP DO LASTPRIVATE(table)
        do ivar=1,size(table)
          table(:ivar) = ivar*5
        end do
    !$OMP END PARALLEL
    if (any(table/=size(table)*5)) then
      stop 'fail 3'
    end if

  !$OMP END PARALLEL
  print *,'ok'
End Program
```
Here's my compilation command -- `flang-new -g -fopenmp -fopenmp-version=31 bug.f90`
When I execute the result, I get the output:
```
Fortran STOP: fail 3
Fortran STOP: fail 3
```

When I compile this with gfortran (`gfortran -fopenmp  bug.f90`) or ifort (`ifort -fopenmp -qopenmp-lib=compat bug.f90`), I get the output:
```
 ok
```

### PR DESCRIPTION
**Pull request by tblah**
The existing conversion inlined private alloc regions and firstprivate copy regions in mlir, then undoing the modification of the mlir module before completing the conversion. To make this work, LLVM IR had to be generated using the wrong mapping for privatised values and then later fixed inside of OpenMPIRBuilder.

This approach violated an assumption in OpenMPIRBuilder that private variables would be values not constants. Flang sometimes generates code where private variables are promoted to globals, the address of which is treated as a constant in LLVM IR. This caused the incorrect values for the private variable from being replaced by OpenMPIRBuilder: ultimately resulting in programs producing incorrect results.

This patch rewrites delayed privatisation for omp.parallel to work more similarly to reductions: translating directly into LLVMIR with correct mappings for private variables.

RFC: https://discourse.llvm.org/t/rfc-openmp-fix-issue-in-mlir-to-llvmir-translation-for-delayed-privatisation/81225

Tested against the gfortran testsuite and our internal test suite. Linaro's post-commit bots will check against the fujitsu test suite.

I decided to add the new tests as flang integration tests rather than in mlir/test/Target/LLVMIR:
  - The regression test is for an issue filed against flang. i wanted to keep the reproducer similar to the code in the ticket.
  - I found the "worst case" CFG test difficult to reason about in abstract it helped me to think about what was going on in terms of a Fortran program.

Fixes #106297

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
The fujitsu testsuite is at https://github.com/fujitsu/compiler-test-suite. Could you give it a try? You can download it into a directory inside llvm-testsuite.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -1405,6 +1375,64 @@ convertOmpParallel(omp::ParallelOp opInst, llvm::IRBuilderBase &builder,
...
+      // in-place convert the private allocation region
+      SmallVector<llvm::Value *, 1> phis;
+      if (allocRegion.getArgument(0).getUses().empty())
+        // TODO this should use
+        // allocaIP.getBlock()->getFirstNonPHIOrDbgOrAlloca() so it goes before
+        // the code for fetching the thread id. Not doing this for now to avoid
+        // test churn.
+        builder.SetInsertPoint(allocaIP.getBlock()->getTerminator());
+      else
+        builder.SetInsertPoint(privAllocBlock->getTerminator());
```
Comment (kiranchandramohan) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1419]: Nit: Add braces

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -1405,6 +1375,64 @@ convertOmpParallel(omp::ParallelOp opInst, llvm::IRBuilderBase &builder,
...
+
+      // map allocation region block argument
+      llvm::Value *llvmArg =
+          moduleTranslation.lookupValue(opInst.getPrivateVars()[i]);
+      assert(llvmArg);
+      moduleTranslation.mapValue(allocRegion.getArgument(0), llvmArg);
+
+      // in-place convert the private allocation region
+      SmallVector<llvm::Value *, 1> phis;
+      if (allocRegion.getArgument(0).getUses().empty())
```
Comment (kiranchandramohan) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1412]: What is this `if` case here? Does it mean that we will not end up using the privAllocaBlock that was created?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 3]
```diff
@@ -1421,12 +1449,56 @@ convertOmpParallel(omp::ParallelOp opInst, llvm::IRBuilderBase &builder,
             deferredStores, isByRef)))
       bodyGenStatus = failure();
 
+    // Apply copy region for firstprivate.
+    if (!privateBlockArgs.empty()) {
+      // Find the end of the allocation blocks
+      assert(afterAllocas->getSinglePredecessor());
+      builder.SetInsertPoint(
+          afterAllocas->getSinglePredecessor()->getTerminator());
+      llvm::BasicBlock *copyBlock = splitBB(builder, true, "omp.private.copy");
```
Comment (kiranchandramohan) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1458]:
```suggestion
      llvm::BasicBlock *copyBlock = splitBB(builder, /*CreateBranch=*/ true, "omp.private.copy");
```

**Comment by tblah** [comment]
> The fujitsu testsuite is at https://github.com/fujitsu/compiler-test-suite. Could you give it a try? You can download it into a directory inside llvm-testsuite.

After my patch I get the same number of "executable missing" tests (which indicate a compilation failure) and 18 **fewer** failed tests

**Review by ergawy — APPROVED** [review]
LGTM, thanks Tom!

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 4]
```diff
@@ -1421,12 +1450,57 @@ convertOmpParallel(omp::ParallelOp opInst, llvm::IRBuilderBase &builder,
             deferredStores, isByRef)))
       bodyGenStatus = failure();
 
+    // Apply copy region for firstprivate.
+    if (!privateBlockArgs.empty()) {
```
[the remark this answers is not in the document]
Comment (tblah) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1454]: I'm only checking that there are private arguments not if any are firstprivate because it felt like overkill to scan every argument looking for a firstprivate one. Would you prefer I did that?

I don't want to always create the blocks because that would lead to unnecessary test churn, especially if they were in OpenMPIRBuilder as that may effect clang too.

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1454]: If we can write a quick std::find_if for firstprivate that would be great.

Could you also have a quick check/thought about whether the move from the separate callback to body gen callback will hamper the transformations that they talk about in the following presentation? 
https://llvm.org/devmtg/2019-04/slides/TechTalk-Doerfert-Compiler_optimization_for_OpenMP_accelerator_offloading.pdf

FYI. No action required. The folks working on llvm openmp were planning on adding and improving the openmp-opt pass. This would work well if the regions produced by both Clang and Flang are similar.
https://openmp.llvm.org/optimizations/OpenMPOpt.html

File: flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90 [hunk 5]
```diff
@@ -0,0 +1,261 @@
+! RUN: %flang_fc1 -fopenmp -emit-llvm %s -o - | FileCheck %s
+
+! stress test cfg and builder insertion points in mlir-to-llvm conversion:
```
Comment (Meinersbur) [inline flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90:3]: I don't think this test matches what's described under https://en.wikipedia.org/wiki/Stress_testing_(computing). For a compiler, I would understand sources that use more than a usual number of function/variable/identifiers, etc to test maybe scaling behavior. Four arrays is not yet an unusual amount. 

I think it could be called combinatorial testing.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 6]
```diff
@@ -371,20 +371,46 @@ convertOmpCritical(Operation &opInst, llvm::IRBuilderBase &builder,
   return success();
 }
 
-/// Populates `reductions` with reduction declarations used in the given loop.
+// Looks up from the operation from and returns the PrivateClauseOp with
+// name symbolName
```
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:375]:
```suggestion
/// Looks up from the operation from and returns the PrivateClauseOp with
/// name symbolName
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 7]
```diff
@@ -371,20 +371,46 @@ convertOmpCritical(Operation &opInst, llvm::IRBuilderBase &builder,
...
+  for (auto symbolRef : attr->getAsRange<SymbolRefAttr>()) {
+    privatizations.push_back(findPrivatizer(op, symbolRef));
+  }
+}
+
+/// Populates `reductions` with reduction declarations used in the given op.
 template <typename T>
 static void
-collectReductionDecls(T loop,
+collectReductionDecls(T op,
```
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:403]: Noise in the diff?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 8]
```diff
@@ -371,20 +371,46 @@ convertOmpCritical(Operation &opInst, llvm::IRBuilderBase &builder,
...
+static omp::PrivateClauseOp findPrivatizer(Operation *from,
+                                           SymbolRefAttr symbolName) {
+  omp::PrivateClauseOp privatizer =
+      SymbolTable::lookupNearestSymbolFrom<omp::PrivateClauseOp>(from,
+                                                                 symbolName);
+  assert(privatizer && "privatizer not found in the symbol table");
+  return privatizer;
+}
+
+/// Populates `privatizations` with privatisation declarations used for the
```
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:385]:
```suggestion
/// Populates `privatizations` with privatization declarations used for the
```

**Review by Meinersbur — COMMENTED** [review]
I like any patch that removes more code than it adds

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tblah) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1454]: Thanks for the links.

> Could you also have a quick check/thought about whether the move from the separate callback to body gen callback will hamper the transformations that they talk about in the following presentation?

Clang doesn't use the separate privatisation callback either (this is equivalent to a no-op in OpenMPIRBuilder): https://github.com/llvm/llvm-project/blob/f3947aaa37f464e05c1a28ce871f5f982c5e2746/clang/lib/CodeGen/CGStmtOpenMP.cpp#L1818

TL;DR: I don't think this PR will make things worse but it would be good to take a closer look at any work needed to enable these optimizations in the future.

My educated guesses on some of the optimizations in the presentation:
- Forwarding communicated/shared values between parallel regions -> I am not sure if this will work for flang but it seems to work using the arguments to the outlined function, which will not be changed by my PR
- Interprocedural constant propagation for outlined parallel regions -> again, the arguments to the outlined function have not changed. The use of the argument should have only changed in the case of the bug I was fixing.
- Attribute deduction also the same as above
- Argument promotion (shared -> firstprivate -> private) -> I don't know how this works. It probably never worked for code generated by flang.
- Parallel region expansion and barrier elimination -> I imagine this should work. It should only need to look at the libcalls generated by OpenMPIRBuilder and maybe stitch some parallel regions together.
- Communication optimization and value transfer -> this should be unaffected by my patch -> it depends on how variables are passed into and out of parallel regions (by ref or by val)

I can't really comment on device code and cross host-device IPO as I'm very unfamiliar with that pipeline and IIUC there's a lot of code yet to be upstreamed.

> FYI. No action required. The folks working on llvm openmp were planning on adding and improving the openmp-opt pass. This would work well if the regions produced by both Clang and Flang are similar.

There are already a few places we diverge (e.g. the reduction init block(s)). Plus the differences that are inherent such as the much higher complexity of Fortran datatypes allowed in privatization and reduction (allocatable arrays etc). Maybe future work could unpack Fortran runtime descriptors in cases where they are not needed so that it looks more like a C array. Currently Flang reductions and privatization force arrays and allocatables to be boxed. 

The runtime call deduplication mentioned on the webpage ought to still work. We still generate runtime calls using OpenMPIRBuilder and the parallel region is still inside of the outlined function.

The webpage also mentions globalization for data shared between different threads on a GPU. I don't know much about the GPU flow.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp (hunk 7 above)
Comment (tblah) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:403]: As I was reading the code I realized this name was misleading because the op might be a ParallelOp. Would you rather I do this in a separate patch?

File: flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90 [hunk 9]
```diff
@@ -0,0 +1,262 @@
+! RUN: %flang_fc1 -fopenmp -emit-llvm %s -o - | FileCheck %s
+
+! Compinational testing of control flow graph and builder insertion points
```
Comment (kiranchandramohan) [inline flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90:3]:
```suggestion
! Combinational testing of control flow graph and builder insertion points
```

File: flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90 [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+! in mlir-to-llvm conversion:
+!   - mixing multiple delayed privatisations and multiple reductions
```
Comment (kiranchandramohan) [inline flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90:5]:
```suggestion
!   - mixing multiple delayed privatizations and multiple reductions
```

File: flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90 [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+!   - multiple blocks in the reduction init region
+!   - reduction init region has to read from the mold variable
+!   - re-used omp.private ops
+!   - re-used omp.reduction.declare ops
+!   - unstructured code inside of the parallel region
+!   - needs private dealloc region, and this has multiple blocks
+!   - needs reduction cleanup region, and this has multiple blocks
+
+! This maybe belongs in the mlir tests, but what we are doing here is complex
+! enough that I find the kind of minimised mlir code prefered by mlir reviwers
```
Comment (kiranchandramohan) [inline flang/test/Integration/OpenMP/parallel-private-reduction-worstcase.f90:19]:
```suggestion
! enough that I find the kind of minimised mlir code preferred by mlir reviewers
```

**Review by kiranchandramohan — APPROVED** [review]
Thanks Tom for the detailed comments and going through the links I gave.

LGTM. Nice work!
